### PR TITLE
Implement progress bar

### PR DESCRIPTION
**Pull request by Razaloc**
Display a progress bar for the copy process

Closes: #33 

Signed-off-by: Rafael Garcia Ruiz <rafael.garcia@collabora.com>

### CODE REVIEW

File: bmap-rs/src/main.rs [hunk 1]
```diff
@@ -2,6 +2,7 @@ use anyhow::{anyhow, bail, Context, Result};
 use bmap::{Bmap, Discarder, SeekForward};
 use clap::Parser;
 use flate2::read::GzDecoder;
+use indicatif::ProgressBar;
```
Comment (obbardc) [inline bmap-rs/src/main.rs:5]: add `Closes: #33` to commit message

File: bmap-rs/src/main.rs [hunk 2]
```diff
@@ -111,7 +120,12 @@ fn copy(c: Copy) -> Result<()> {
     }
 
     let mut input = setup_input(&c.image)?;
-    bmap::copy(&mut input, &mut output, &bmap)?;
+    if custom.pbar {
+        let pb = ProgressBar::new(bmap.image_size());
+        bmap::copy(&mut input, &mut pb.wrap_write(&output), &bmap)?;
```
[the remark this answers is not in the document]
Comment (Razaloc) [reply bmap-rs/src/main.rs:125]: Exactly! anyway I've been experimenting with callbacks and closures

File: bmap-rs/src/main.rs [hunk 3]
```diff
@@ -111,7 +120,12 @@ fn copy(c: Copy) -> Result<()> {
     }
 
     let mut input = setup_input(&c.image)?;
-    bmap::copy(&mut input, &mut output, &bmap)?;
+    if custom.pbar {
+        let pb = ProgressBar::new(bmap.image_size());
```
[the remark this answers is not in the document]
Comment (obbardc) [reply bmap-rs/src/main.rs:124]: Also, using image_size is wrong for the compressed image size.

We should have a helper function under `bmap` to calculate `mapped_image_size` at parsetime which would calculate `BlockSize * MappedBlocksCount`
```
    <!-- Image size in bytes: 16.0 MiB -->
    <ImageSize> 16777216 </ImageSize>

    <!-- Size of a block in bytes -->
    <BlockSize> 4096 </BlockSize>

    <!-- Count of blocks in the image file -->
    <BlocksCount> 4096 </BlocksCount>

    <!-- Count of mapped blocks: 4.0 MiB or 25.0%     -->
    <MappedBlocksCount> 1026 </MappedBlocksCount>
```

[the remark this answers is not in the document]
Comment (Razaloc) [reply bmap-rs/src/main.rs]: I don't understand. Ain't <ImageSize> field just the same? Bmap file is done from the uncompressed file so I was certain that compressed or not the value should be valid.

[the remark this answers is not in the document]
Comment (Razaloc) [reply bmap-rs/src/main.rs]: I mean, as I saw it compression is unrelated to the bmap file

[the remark this answers is not in the document]
Comment (Razaloc) [reply bmap-rs/src/main.rs]: I see, thanks

File: bmap/src/bmap.rs [hunk 4]
```diff
@@ -92,6 +92,9 @@ impl Bmap {
     pub fn block_map(&self) -> impl ExactSizeIterator + Iterator<Item = &BlockRange> {
         self.blockmap.iter()
     }
+    pub fn total_size(&self) -> u64 {
+        self.block_size * self.mapped_blocks
```
Comment (obbardc) [inline bmap/src/bmap.rs:96]: call it total_mapped_size

File: bmap-rs/src/main.rs [hunk 5]
```diff
@@ -111,7 +113,14 @@ fn copy(c: Copy) -> Result<()> {
     }
 
     let mut input = setup_input(&c.image)?;
-    bmap::copy(&mut input, &mut output, &bmap)?;
+    let pb = ProgressBar::new(bmap.total_size());
+    pb.set_style(ProgressStyle::with_template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")
+        .unwrap()
+        .with_key("eta", |state: &ProgressState, w: &mut dyn Write| write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap())
+        .progress_chars("#>-"));
+    bmap::copy(&mut input, &mut pb.wrap_write(&output), &bmap)?;
```
Comment (obbardc) [inline bmap-rs/src/main.rs:121]: Quick question, does this easily work with the other PR which implements the async case ?

Comment (Razaloc) [reply bmap-rs/src/main.rs:121]: there is a async version of the wrap

**Comment by obbardc** [comment]
this looks really nice; great work @Razaloc 

bors r+

**Pull request merged**